Add support for #derive[FromStr] for enums

aj-bagwell opened the pull request:
This allows deriving FromStr for enums with no fields in the obvious way

fixes #59

aj-bagwell commented on src/from_str.rs:
```diff
@@ -41,7 +52,81 @@ pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStre
+            type Err = #err_name;
+            #[inline]
+            fn from_str(src: &str) -> ::core::result::Result<Self, Self::Err> {
+                Ok(match src.to_lowercase().as_str() {
```
I considered adding `.replace(&[' ', '_', '-'], "")` here so that it would mach against kebab-case and SNAKE_CASE too.

aj-bagwell commented on src/from_str.rs:
```diff
@@ -41,7 +52,81 @@ pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStre
+    quote! {
+
+        #[derive(Debug, Clone, PartialEq, Eq)]
+        #visibility struct #err_name;
```
I'm not sure I like creating a type out of nothing here, but I did not like the other options better:

- could use `&'static str` but that does not impelement `std::error::Error` which make it annoying for combining with other errors
- could have a `derive_more::ParseError` but that makes the dependencies complicates as you cant have code and procmarcos in the same dependency

aj-bagwell commented:
Forced push changes that hopefull fix the failing tests

JelteF commented on Cargo.toml:
```diff
@@ -51,7 +51,7 @@ deref_mut = []
 error = ["syn/extra-traits"]
 from = ["syn/extra-traits"]
-from_str = []
+from_str = ["convert_case"]
```
This should fix the current compilation issue in the feature tests.
```suggestion
from_str = ["convert_case", "syn/extra-traits"]
```

aj-bagwell replied:
Thanks, I just furoced pushed a change that bumps `syn` to 1.0.81 (which is the version in the lock file) and changed the code to use .is_empty() instead which means it does not need the extra traits

aj-bagwell replied:
Bah looks like I also need to ditch the `x == from_str(x.to_string())` test as I can't `#derive[Display]` if the display feature is not set.

aj-bagwell commented:
The test are all now passing and I'm happy with my commits as they stand, @JelteF let me know if there are any more changes that you want me to make.

JelteF commented on src/from_str.rs:
```diff
@@ -41,7 +51,81 @@ pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStre
+                #canonical => #input_type::#variant,
+            })
+        } else {
+            for variant in variants {
+                let variant_str = variant.to_string();
+                cases.push(quote! {
+                    #canonical if(src == #variant_str) => #input_type::#variant,
+                })
+            }
```
this edge case is missing a test

JelteF reviewed (dismissed):
I think this is good except for a missing test for the smart case sensitive behaviour.

aj-bagwell commented:
I have added `enum_test_case_sensitive` to check if falls back to case sentive matching when it matters and rebased onto master to bring it up to date

JelteF commented:
Thanks a lot for your work on this!
